### PR TITLE
[FIX] sale_timesheet: Fix error when deleting the 'Service on Timesheet' product

### PR DESCRIPTION
**Pull request by SimonGenin**
This commit prevents a user from deleting or archiving the default
"Service on Timesheet" product.

Before this commit, the users were receiving a constraint violation
error.

website_sale test TestWebsiteSequence set up had to take this change
into account. A force argument to the product.write function has been
added to force the archiving of the product.

TaskID: 2221046

### CODE REVIEW

File: addons/sale_timesheet/models/product.py [hunk 1]
```diff
@@ -19,6 +20,12 @@ class ProductTemplate(models.Model):
     project_id = fields.Many2one(domain="[('billable_type', '=', 'no'), ('allow_timesheets', 'in', [service_policy == 'delivered_timesheet' or '', True])]")
     project_template_id = fields.Many2one(domain="[('billable_type', '=', 'no'), ('allow_timesheets', 'in', [service_policy == 'delivered_timesheet' or '', True])]")
 
+    def unlink(self):
+        time_product = self.env.ref('sale_timesheet.time_product')
+        if time_product.product_tmpl_id.id in self.ids:
```
Comment (fmdl) [inline addons/sale_timesheet/models/product.py:25]: `if time_product.product_tmpl_id in self:`

File: addons/website_sale/tests/test_website_sequence.py [hunk 2]
```diff
@@ -11,10 +11,12 @@ def setUp(self):
 
         ProductTemplate = self.env['product.template']
         product_templates = ProductTemplate.search([])
+
```
Comment (kba-odoo) [inline addons/website_sale/tests/test_website_sequence.py:14]: You can remove those 2 extra lines, they are not needed here :-)

File: addons/sale_timesheet/models/product.py [hunk 3]
```diff
@@ -19,6 +20,12 @@ class ProductTemplate(models.Model):
     project_id = fields.Many2one(domain="[('billable_type', '=', 'no'), ('allow_timesheets', 'in', [service_policy == 'delivered_timesheet' or '', True])]")
     project_template_id = fields.Many2one(domain="[('billable_type', '=', 'no'), ('allow_timesheets', 'in', [service_policy == 'delivered_timesheet' or '', True])]")
 
+    def unlink(self):
+        time_product = self.env.ref('sale_timesheet.time_product')
+        if time_product.product_tmpl_id in self:
+            raise ValidationError(_('This product cannot be archived, nor deleted.'))
```
Comment (mart-e) [inline addons/sale_timesheet/models/product.py:26]: could be good to actually say which product is problematic (think if you remove 80 products)

**Comment by tivisse** [comment]
Superseded by https://github.com/odoo/enterprise/pull/11077

File: addons/sale_timesheet/models/product.py (hunk 3 above)
Comment (tivisse) [reply addons/sale_timesheet/models/product.py:26]: Good point ;)

**Pull request closed without merging**